### ISSUE TITLE
Are changes to the System seen by the Context? Documentation and results disagree...

### ISSUE THREAD

**Issue by apayne97** [openmm/openmm#3345]
thanks to @zhang-ivy for the simplified code examples here. Simple question is, if we add a custom force to the system after creating the simulation context, will the context see the change? The [documentation](http://docs.openmm.org/latest/api-python/generated/openmm.openmm.Context.html?highlight=context#openmm.openmm.Context.reinitialize) says no. But this code disagrees:

First we test if changing the system changes the context. It does.
```
from openmmtools.integrators import LangevinIntegrator
import simtk.openmm as mm
from simtk.openmm import unit

system = mm.System()
system.addParticle(4 * unit.amu)
integrator = LangevinIntegrator(temperature=300 * unit.kelvin, 
                                collision_rate=1 / unit.picoseconds, 
                                timestep=4 * unit.femtoseconds)
context = mm.Context(system, integrator)
print(context.getSystem().getForces()) # This will not contain any forces
system.addForce(mm.HarmonicBondForce())
print(context.getSystem().getForces()) # If the reinitialize() docs are correct, this should not return any forces
```

```
[]
[<simtk.openmm.openmm.HarmonicBondForce; proxy of <Swig Object of type 'OpenMM::HarmonicBondForce *' at 0x7f8573432de0> >]
```

Then we test if changing the context changes the saved system. It does.
```
system = mm.System()
system.addParticle(4 * unit.amu)
integrator = LangevinIntegrator(temperature=300 * unit.kelvin, 
                                collision_rate=1 / unit.picoseconds, 
                                timestep=4 * unit.femtoseconds)
context = mm.Context(system, integrator)
print(context.getSystem().getForces()) # This will not contain any forces
context.getSystem().addForce(mm.HarmonicBondForce())
print(system.getForces()) # If the reinitialize() docs are correct, this should not return any forces
```

I understand why this happens. And honestly it's convenient to be able to change the system after creating the context. But it seems like the documentation and behaviour should agree. And if it is true that `.getForces()` doesn't actually return a true representation of the Context, that also seems like a problem. 

Thanks!

**Comment by peastman** [comment]
The documentation is correct.  The context does not see any changes you make to the system.  By printing `system.getForces()`, all you're showing is that the system has been modified, which of course it has.  You added another force to it.  But the context isn't aware of the existence of that force.  That means, for example,

- When forces and energies are evaluated, there will be no contribution from the new force.
- If the force defines an `updateContextState()` method to modify the context (for example, MonteCarloBarostat or AndersenThermostat), that method will never get called.
- If the new force defines any parameters, they won't show up in `context.getParameters()`.

Your simulation will continue to run as if the new force didn't exist.  If you want it to start affecting the simulation, you need to call `reinitialize()`.

**Comment by apayne97** [comment]
So does `context.getSystem()` not return the true system? because in the first example we print `context.getSystem().getForces()` and it has been changed.

**Comment by peastman** [comment]
There's only one System object involved.  When you create a Context, it doesn't clone the System, it just stores a reference to the existing object.  In your code, `system` and `context.getSystem()` are the same object.  But the Context is not aware of any changes you make to that object unless you reinitialize it.

**Comment by jchodera** [comment]
@apayne97 : Does this answer your question? If so, is further documentation clarification needed to make these points clearer?

**Comment by apayne97** [comment]
Yes, it answers the question. I understand what is going on. But to me it seems like confusing behaviour. I would have expected that either 1) the simulation __init__ would create a new system object or that 2) the simulation would just save a pointer to the simulation object and would reinitialize before starting the simulation in order to enable arbitrary changes to the system object. In either case,  and therefore whenever you call `context.getSystem()` would accurately represent the system that will be used during the simulation. But in this case, you can't trust that the system you get from `context.getSystem()` is the one that will actually be used during the simulation.

Given that both reinitializing and saving a new system object are computationally costly (?), I see why it works like this. But a way to make this clearer either in the Simulation documentation or even as a part of the `context.getSystem()` code would be good, I think. Maybe even just a warning when you run `context.getSystem()` that if you have changed the system the context won't see it unless you reinitialize?

Idk. Maybe most people will avoid encountering this. But since Ivy and I both did, it might be worth doing something to make it more obvious?

**Comment by jchodera** [comment]
It definitely sounds like we could add some eye-popping warnings to both the User Guide and API docs to emphasize this behavior, since I agree it is "surprising" (one of the least desirable characteristics of any API). 

Perhaps we could rename the issue to "Update documentation to emphasize changes to System objects retrieved from Context are not automatically updated", or something that more clearly captures what we need to fix?

The only other minimally-intrusive behavior that might make sense here is to consider returning a _copy_ of the `System` object when `Context.getSystem()` is called. It is indeed much more surprising if changes one makes persist even though the context is not updated, especially because the process for updating the system in the `Context` (`Force.updateParametersInContext(context)`) does not care whether the `System` object is owned by the `Context` or not. But I'm not sure if this would have other effects.

**Comment by peastman** [comment]
> The only other minimally-intrusive behavior that might make sense here is to consider returning a copy of the System object when Context.getSystem() is called.

That would break compatibility in major ways.  It would also make it impossible to get the System, modify it, and then call `reinitialize()` or `updateParametersInContext()`.
